core/src/main/java/org/apache/accumulo/core/util/ServiceLockData.java

```diff
@@ -0,0 +1,224 @@
+      return toString().hashCode();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (o instanceof ServiceDescriptor) {
+        return toString().equals(o.toString());
+      }
+      return false;
+    }
+
+    @Override
+    public String toString() {
+      return gson.toJson(this);
+    }
+
+  }
+
+  /**
+   * A set of ServiceDescriptor's
+   */
+  public static class ServiceDescriptors {
+    private final Set<ServiceDescriptor> descriptors;
+
+    public ServiceDescriptors() {
+      descriptors = new HashSet<>();
+    }
+
+    public ServiceDescriptors(HashSet<ServiceDescriptor> descriptors) {
+      this.descriptors = descriptors;
+    }
+
+    public void addService(ServiceDescriptor sd) {
+      this.descriptors.add(sd);
+    }
+
+    public Set<ServiceDescriptor> getServices() {
+      return descriptors;
+    }
+  }
+
+  private EnumMap<ThriftService,ServiceDescriptor> services;
+
+  public ServiceLockData(ServiceDescriptors sds) {
+    this.services = new EnumMap<>(ThriftService.class);
+    sds.getServices().forEach(sd -> {
+      this.services.put(sd.getService(), sd);
+    });
+  }
+
+  public ServiceLockData(UUID uuid, String address, ThriftService service, String group) {
+    this(new ServiceDescriptors(new HashSet<>(
+        Collections.singleton(new ServiceDescriptor(uuid, service, address, group)))));
+  }
+
+  public ServiceLockData(UUID uuid, String address, ThriftService service) {
+    this(new ServiceDescriptors(
+        new HashSet<>(Collections.singleton(new ServiceDescriptor(uuid, service, address)))));
+  }
+
+  public String getAddressString(ThriftService service) {
+    ServiceDescriptor sd = services.get(service);
+    if (sd == null) {
+      return null;
+    }
+    return sd.getAddress();
+  }
+
+  public HostAndPort getAddress(ThriftService service) {
+    return AddressUtil.parseAddress(getAddressString(service), false);
+  }
+
+  public String getGroup(ThriftService service) {
+    ServiceDescriptor sd = services.get(service);
+    if (sd == null) {
+      return null;
+    }
+    return sd.getGroup();
+  }
+
+  public UUID getServerUUID(ThriftService service) {
+    ServiceDescriptor sd = services.get(service);
+    if (sd == null) {
+      return null;
+    }
+    return sd.getUUID();
+  }
+
+  public String serialize() {
+    ServiceDescriptors sd = new ServiceDescriptors();
+    services.values().forEach(s -> sd.addService(s));
+    return gson.toJson(sd);
+  }
+
+  @Override
+  public String toString() {
+    return serialize();
+  }
+
+  @Override
+  public int hashCode() {
+    return toString().hashCode();
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (o instanceof ServiceLockData) {
+      return toString().equals(o.toString());
+    }
+    return false;
+  }
+
+  @Override
+  public int compareTo(ServiceLockData other) {
+    return toString().compareTo(other.toString());
+  }
+
+  public static Optional<ServiceLockData> parse(String lockData) {
```

Comment:
Would this be more general if it took `byte[]` and did the string conversion internally to the method?